pkg/network/ebpf/c/protocols/tls/tls.h

```diff
@@ -1,132 +1,459 @@
+static __always_inline bool parse_supported_versions_extension(struct __sk_buff *skb, __u64 *offset, __u32 data_end, __u64 extensions_end, tls_info_t *tags, bool is_client_hello) {
+    if (is_client_hello) {
+        // Read supported version list length (1 byte)
+        if (*offset + SINGLE_BYTE_LENGTH > data_end || *offset + SINGLE_BYTE_LENGTH > extensions_end) {
+            return false;
+        }
+        __u8 sv_list_length;
+        if (bpf_skb_load_bytes(skb, *offset, &sv_list_length, SINGLE_BYTE_LENGTH) < 0) {
+            return false;
+        }
+        *offset += SINGLE_BYTE_LENGTH;
+
+        if (*offset + sv_list_length > data_end || *offset + sv_list_length > extensions_end) {
+            return false;
+        }
+
+        // Parse the list of supported versions (2 bytes each)
+        __u8 sv_offset = 0;
+        __u16 sv_version;
+        #pragma unroll(MAX_SUPPORTED_VERSIONS)
+        for (int idx = 0; idx < MAX_SUPPORTED_VERSIONS; idx++) {
+            if (sv_offset + 1 >= sv_list_length) {
+                break;
+            }
+            // Each supported version is 2 bytes
+            if (*offset + PROTOCOL_VERSION_LENGTH > data_end) {
+                return false;
+            }
+
+            if (bpf_skb_load_bytes(skb, *offset, &sv_version, PROTOCOL_VERSION_LENGTH) < 0) {
+                return false;
+            }
+            sv_version = bpf_ntohs(sv_version);
+            *offset += PROTOCOL_VERSION_LENGTH;
+
+            set_tls_offered_version(tags, sv_version);
+            sv_offset += PROTOCOL_VERSION_LENGTH;
+        }
+    } else {
+        // ServerHello
+        // The selected_version field is 2 bytes
+        if (*offset + PROTOCOL_VERSION_LENGTH > data_end) {
+            return false;
+        }
+
+        // Read selected version (2 bytes)
+        __u16 selected_version;
+        if (bpf_skb_load_bytes(skb, *offset, &selected_version, PROTOCOL_VERSION_LENGTH) < 0) {
+            return false;
+        }
+        selected_version = bpf_ntohs(selected_version);
+        *offset += PROTOCOL_VERSION_LENGTH;
+
+        tags->chosen_version = selected_version;
+    }
+
+    return true;
 }
 
-// is_tls_handshake checks if the given TLS message header is a valid TLS
-// handshake message. The message is considered valid if:
-// - The type matches CLIENT_HELLO or SERVER_HELLO
-// - The version is a known SSL/TLS version
-static __always_inline bool is_tls_handshake(tls_record_header_t *hdr, const char *buf, __u32 buf_size) {
-    // Checking the buffer size contains at least the size of the tls record header and the tls hello message header.
-    if (sizeof(tls_record_header_t) + sizeof(tls_hello_message_t) > buf_size) {
+// parse_tls_extensions parses TLS extensions and looks for the supported_versions extension
+static __always_inline bool parse_tls_extensions(struct __sk_buff *skb, __u64 *offset, __u32 data_end, __u64 extensions_end, tls_info_t *tags, bool is_client_hello) {
```

Comment:
please add a link to the RFC structs message so people can refer to the decoding logic here